src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,296 @@
+                for equivalence in equivalences.iter() {
+                    for expr in equivalence {
+                        needed.extend(expr.support());
+                    }
+                }
+                let mut col_map = HashMap::new();
+
+                let mut total_arity = 0;
+
+                for (idx, input) in inputs.iter_mut().enumerate() {
+                    let mut proj = vec![];
+                    for col in 0..input.arity() {
+                        let global_col = input_mapper.map_column_to_global(col, idx);
+                        if needed.contains(&global_col) {
+                            col_map.insert(global_col, total_arity);
+                            total_arity += 1;
+                            proj.push(col);
+                        }
+                    }
+                    self.action(input, proj);
+                }
+
+                for equivs in equivalences.iter_mut() {
+                    for expr in equivs.iter_mut() {
+                        expr.visit_mut(&mut |e| {
+                            if let ScalarExpr::Column(c) = e {
+                                *c = *col_map.get(c).unwrap();
+                            }
+                        });
+                    }
+                }
+
+                *relation = RelationExpr::join_scalars(inputs.to_vec(), equivalences.clone())
+                    .project(
+                        projection
+                            .into_iter()
+                            .map(|c| *col_map.get(&c).unwrap())
+                            .collect(),
+                    );
+            }
+            RelationExpr::Reduce {
+                input,
+                group_key,
+                aggregates,
+                monotonic: _,
+            } => {
+                // TODO(justin); do we need to preserve `monotonic` here?
+                let input_arity = input.arity();
+                let needed: HashSet<_> = projection
+                    .iter()
+                    .flat_map(|p| {
+                        if *p < group_key.len() {
+                            group_key[*p].support()
+                        } else {
+                            aggregates[*p - group_key.len()].expr.support()
+                        }
+                    })
+                    .collect();
+
+                let mut proj = vec![];
+                let mut input_col_map = HashMap::new();
+
+                for i in 0..input_arity {
+                    if needed.contains(&i) {
+                        input_col_map.insert(i, proj.len());
+                        proj.push(i);
+                    }
+                }
+
+                for a in aggregates.iter_mut() {
+                    a.expr.visit_mut(&mut |e| {
+                        if let ScalarExpr::Column(c) = e {
+                            *c = *input_col_map.get(c).unwrap();
+                        }
+                    });
```

Comment:
This pattern recurs a lot. There is a method `ScalarExpr::permute` that goes in the other direction (performs a vector lookup, going from dense identifiers to sparse identifiers). Perhaps we should create a similar method to rewrite expressions for an arbitrary column remapping?

Reply:
PR incoming.